In [14]:
import sys
import pandas as pd

sys.path.append('../utils')
from plot_func_notebook import create_filter_plot, plot_categorical_bar_chart, plot_ratio_by_configurations, filter_wall_df
from utils_func_notebook import fit_data_to_distance, fit_data_to_distance_thickness

# Charger vos données
data = pd.read_excel("../DB/All-at-once_DB.xlsx", sheet_name='wall')

FileNotFoundError: [Errno 2] No such file or directory: '../DB/All-at-once_DB.xlsx'

In [2]:
create_filter_plot(data)

Output()

In [6]:
particles = ['N']  # ['N', 'P']  
screens = ['Lead'] # ['Concrete', 'Steel', 'Water', 'Lead']
cases = ['C1 [U(4.95)O2F2 (H/235U = 410)]']  # ['C1 [U(4.95)O2F2 (H/235U = 410)]', 'C4 [U(93.2) (H/235U = 0)]']
codes = ['MCNP 6.1']  #  ['MCNP 6.1', 'COG 11.2', 'SCALE 6.2.2']
thicknesses = [] # [1, 5, 10, 20]

In [4]:
particles = ['P']  # ['N', 'P']  
screens = ['Lead'] # ['Concrete', 'Steel', 'Water', 'Lead']
cases = ['C1 [U(4.95)O2F2 (H/235U = 410)]']  # ['C1 [U(4.95)O2F2 (H/235U = 410)]', 'C4 [U(93.2) (H/235U = 0)]']
codes = ['MCNP 6.1']  #  ['MCNP 6.1', 'COG 11.2', 'SCALE 6.2.2']
thicknesses = [5, 10, 20] # [1, 5, 10, 20]

In [7]:
plot_ratio_by_configurations(data, particles, screens, cases, codes, thicknesses)

In [6]:
plot_categorical_bar_chart(data, particles, screens, cases, codes, thicknesses)

____

In [ ]:
import sys
import pandas as pd

sys.path.append('../utils')
from plot_func_notebook import create_filter_plot, plot_categorical_bar_chart, plot_ratio_by_configurations, filter_wall_df
from utils_func_notebook import fit_data_to_distance, fit_data_to_distance_thickness

# Charger vos données
data = pd.read_excel("../DB/All-at-once_DB.xlsx", sheet_name='wall')

In [ ]:
# Define your filter criteria
particles = ['N']  # Example: ['N', 'P']
screens = ['None']  # Example: ['None','Concrete', 'Steel', 'Water', 'Lead']
cases = ['C4 [U(93.2) (H/235U = 0)]']  # Add more cases as needed
codes = ['MCNP 6.1']  # Example: ['MCNP 6.1', 'COG 11.2', 'SCALE 6.2.2']
thicknesses = [0]  # Example: [1, 5, 10, 20, 40]
# Apply the filter and plot the data
filtered_data = filter_wall_df(data, particles, screens, cases, codes, thicknesses, plot=True)

# Display the filtered data
filtered_data

,Fissile,Case,Code,Library,Flux-to-dose conversion factor,Screen,Thickness (cm),Particle,Distance (m),Dose (Gy),1s uncertainty,Absolute Uncertainty,Filter Combo
2210,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,1,3.406024e+01,0.0008,2.724819e-02,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2211,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,2,1.034946e+01,0.0012,1.241935e-02,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2212,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,5,1.875706e+00,0.0019,3.563842e-03,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2213,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,10,4.748761e-01,0.0029,1.377141e-03,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2214,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,20,1.196466e-01,0.0044,5.264448e-04,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2215,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,50,2.011397e-02,0.0076,1.528662e-04,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2216,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,100,5.030826e-03,0.0106,5.332676e-05,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2217,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,200,1.022377e-03,0.0110,1.124615e-05,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2218,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,300,3.150162e-04,0.0104,3.276169e-06,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0
2219,U,C4 [U(93.2) (H/235U = 0)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,None,0,N,500,4.704534e-05,0.0102,4.798624e-07,N_None_MCNP 6.1_C4 [U(93.2) (H/235U = 0)]_0


---

# 📊 Exploration : Tracer en fonction d'autres variables

Testons la possibilité de tracer les résultats en fonction de différentes variables (épaisseur, densité, etc.)

In [5]:
# Charger les données et explorer les colonnes disponibles
import pandas as pd

data_wall = pd.read_excel("../Database/All-at-once_DB.xlsx", sheet_name='wall')
print("Colonnes disponibles dans la base de données:")
print(data_wall.columns.tolist())
print("\n📏 Premières lignes:")
data_wall.head()

Colonnes disponibles dans la base de données:
['Fissile', 'Case', 'Code', 'Library', 'Flux-to-dose conversion factor', 'Screen', 'Thickness (cm)', 'Particle', 'Distance (m)', 'Dose (Gy)', '1s uncertainty']

📏 Premières lignes:


,Fissile,Case,Code,Library,Flux-to-dose conversion factor,Screen,Thickness (cm),Particle,Distance (m),Dose (Gy),1s uncertainty
0,U,C1 [U(4.95)O2F2 (H/235U = 410)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,Concrete,20,N,1,2.457272,0.0225
1,U,C1 [U(4.95)O2F2 (H/235U = 410)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,Concrete,20,N,2,0.772401,0.0084
2,U,C1 [U(4.95)O2F2 (H/235U = 410)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,Concrete,20,N,5,0.143430,0.0097
3,U,C1 [U(4.95)O2F2 (H/235U = 410)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,Concrete,20,N,10,0.035122,0.0093
4,U,C1 [U(4.95)O2F2 (H/235U = 410)],MCNP 6.1,ENDF/B-VII.1,ANSI/HPS N13.3,Concrete,20,N,20,0.008329,0.0071


In [6]:
import plotly.graph_objects as go
import plotly.express as px
import numpy as np

def plot_dose_vs_variable(data, x_variable='Distance (m)', filters=None):
    """
    Trace la dose en fonction d'une variable au choix.
    
    Parameters:
    -----------
    data : DataFrame
        Les données à tracer
    x_variable : str
        La variable à mettre en abscisse (par défaut 'Distance (m)')
        Options: 'Distance (m)', 'Thickness (cm)', etc.
    filters : dict
        Dictionnaire de filtres à appliquer
        Exemple: {'Particle': ['N'], 'Screen': ['Lead'], 'Case': [...]}
    """
    
    # Copier les données pour ne pas modifier l'original
    df = data.copy()
    
    # Appliquer les filtres si fournis
    if filters:
        for column, values in filters.items():
            if values:  # Si la liste n'est pas vide
                df = df[df[column].isin(values)]
    
    # Créer le graphique
    fig = go.Figure()
    
    # Grouper par toutes les variables sauf x_variable et Dose
    group_columns = [col for col in ['Fissile', 'Case', 'Code', 'Screen', 'Thickness (cm)', 'Particle', 'Distance (m)'] 
                     if col != x_variable and col in df.columns]
    
    # Créer une série pour chaque combinaison unique
    for name, group in df.groupby(group_columns):
        # Créer un label descriptif
        if isinstance(name, tuple):
            label_parts = [f"{col}={val}" for col, val in zip(group_columns, name)]
        else:
            label_parts = [f"{group_columns[0]}={name}"]
        
        label = ", ".join(label_parts)
        
        # Trier par x_variable
        group_sorted = group.sort_values(x_variable)
        
        # Ajouter la trace
        fig.add_trace(go.Scatter(
            x=group_sorted[x_variable],
            y=group_sorted['Dose'],
            mode='lines+markers',
            name=label,
            error_y=dict(
                type='data',
                array=group_sorted['1s uncertainty'] * group_sorted['Dose'],
                visible=True
            )
        ))
    
    # Mise en forme avec légende en dessous
    fig.update_layout(
        title=f"Dose en fonction de {x_variable}",
        xaxis_title=x_variable,
        yaxis_title="Dose (Gy)",
        yaxis_type="log",
        hovermode='closest',
        height=600,
        showlegend=True,
        legend=dict(
            orientation="h",
            yanchor="top",
            y=-0.2,
            xanchor="center",
            x=0.5
        )
    )
    
    return fig

# Exemple d'utilisation
print("✅ Fonction plot_dose_vs_variable créée !")
print("\nVariables disponibles pour l'axe X:")
print("- 'Distance (m)' (par défaut)")
print("- 'Thickness (cm)'")
print("- Ou toute autre colonne numérique")

✅ Fonction plot_dose_vs_variable créée !

Variables disponibles pour l'axe X:
- 'Distance (m)' (par défaut)
- 'Thickness (cm)'
- Ou toute autre colonne numérique


## Test 1 : Dose en fonction de l'épaisseur d'écran

Pour un cas donné, voyons comment la dose varie avec l'épaisseur de l'écran à différentes distances

In [10]:
# Tracer la dose en fonction de l'épaisseur pour un cas spécifique
filters = {
    'Particle': ['N'],
    'Screen': ['Lead'],
    'Case': ['C1 [U(4.95)O2F2 (H/235U = 410)]'],
    'Code': ['MCNP 6.1'],
    # Pas de filtre sur Distance pour afficher toutes les distances disponibles
}

fig = plot_dose_vs_variable(data_wall, x_variable='Thickness (cm)', filters=filters)
fig.show()

## Test 2 : Comparaison de différents écrans en fonction de l'épaisseur

À distance fixe, comparons l'efficacité de différents matériaux d'écran

In [8]:
# Comparer différents matériaux à distance fixe
filters = {
    'Particle': ['N'],
    'Screen': ['Concrete', 'Steel', 'Lead', 'Water'],  # Tous les matériaux
    'Case': ['C1 [U(4.95)O2F2 (H/235U = 410)]'],
    'Code': ['MCNP 6.1'],
    'Distance (m)': [5]  # Distance fixe
}

fig = plot_dose_vs_variable(data_wall, x_variable='Thickness (cm)', filters=filters)
fig.show()

## Test 3 : Graphique traditionnel - Dose en fonction de la distance

Pour comparaison, voici le graphique classique avec différentes épaisseurs d'écran

In [9]:
# Graphique classique : dose vs distance pour différentes épaisseurs
filters = {
    'Particle': ['N'],
    'Screen': ['Lead'],
    'Case': ['C1 [U(4.95)O2F2 (H/235U = 410)]'],
    'Code': ['MCNP 6.1'],
    'Thickness (cm)': [1, 5, 10, 20, 40]
}

fig = plot_dose_vs_variable(data_wall, x_variable='Distance (m)', filters=filters)
fig.show()

## 📝 Observations et conclusions

### Ce que nous pouvons voir avec ces nouveaux graphiques :

**Test 1** - Dose vs Épaisseur (à différentes distances) :
- Permet de voir comment l'efficacité de l'écran évolue avec son épaisseur
- Montre clairement l'atténuation exponentielle avec l'épaisseur
- Utile pour dimensionner un écran à une distance donnée

**Test 2** - Comparaison des matériaux (épaisseur variable, distance fixe) :
- Compare directement l'efficacité des différents matériaux
- Permet de choisir le meilleur matériau pour une application donnée
- Montre les coefficients d'atténuation relatifs

**Test 3** - Graphique classique (distance variable, épaisseurs fixes) :
- Vue traditionnelle de l'atténuation avec la distance
- Montre l'effet combiné de la géométrie (1/r²) et de l'atténuation

### 💡 Pour intégrer dans l'application Streamlit :

Il faudrait ajouter un **sélecteur d'axe X** dans l'interface qui permettrait de choisir :
- Distance (m) → graphique actuel
- Thickness (cm) → nouveau graphique
- Éventuellement d'autres variables si pertinent

Cela nécessiterait :
1. Ajouter un `st.selectbox()` pour choisir la variable X
2. Adapter la fonction de plotting pour supporter différentes variables
3. Ajuster les filtres en conséquence (ne pas filtrer sur la variable qu'on trace)

## 🔄 Test 4 : Facteur d'atténuation normalisé (dose/dose sans écran)

Normalisation par rapport à la dose sans écran pour voir directement l'efficacité de l'écran

In [11]:
def plot_normalized_dose(data, x_variable='Thickness (cm)', normalize_by='Thickness (cm)', 
                         reference_value=0, filters=None):
    """
    Trace la dose normalisée par rapport à une valeur de référence.
    
    Parameters:
    -----------
    data : DataFrame
        Les données à tracer
    x_variable : str
        La variable à mettre en abscisse
    normalize_by : str
        La colonne utilisée pour la normalisation
    reference_value : float
        La valeur de référence pour la normalisation (ex: 0 pour épaisseur sans écran)
    filters : dict
        Dictionnaire de filtres à appliquer
    """
    
    # Copier les données
    df = data.copy()
    
    # Appliquer les filtres
    if filters:
        for column, values in filters.items():
            if values:
                df = df[df[column].isin(values)]
    
    # Créer le graphique
    fig = go.Figure()
    
    # Grouper par toutes les variables sauf x_variable et normalize_by
    group_columns = [col for col in ['Fissile', 'Case', 'Code', 'Screen', 'Particle', 'Distance (m)'] 
                     if col != x_variable and col != normalize_by and col in df.columns]
    
    # Pour chaque groupe
    for name, group in df.groupby(group_columns):
        # Créer un label descriptif
        if isinstance(name, tuple):
            label_parts = [f"{col}={val}" for col, val in zip(group_columns, name)]
        else:
            label_parts = [f"{group_columns[0]}={name}"]
        
        label = ", ".join(label_parts)
        
        # Trouver la dose de référence (sans écran)
        ref_dose = group[group[normalize_by] == reference_value]['Dose']
        
        if len(ref_dose) == 0:
            print(f"⚠️ Pas de valeur de référence ({normalize_by}={reference_value}) pour {label}")
            continue
        
        ref_dose_value = ref_dose.iloc[0]
        
        # Normaliser les doses
        group_sorted = group.sort_values(x_variable).copy()
        group_sorted['Normalized Dose'] = group_sorted['Dose'] / ref_dose_value
        
        # Propager les incertitudes (division par une constante)
        group_sorted['Normalized Uncertainty'] = group_sorted['1s uncertainty']
        
        # Ajouter la trace
        fig.add_trace(go.Scatter(
            x=group_sorted[x_variable],
            y=group_sorted['Normalized Dose'],
            mode='lines+markers',
            name=label,
            error_y=dict(
                type='data',
                array=group_sorted['Normalized Uncertainty'] * group_sorted['Normalized Dose'],
                visible=True
            )
        ))
    
    # Mise en forme
    fig.update_layout(
        title=f"Facteur d'atténuation en fonction de {x_variable}<br>(normalisé par dose sans écran)",
        xaxis_title=x_variable,
        yaxis_title=f"Dose / Dose(sans écran)",
        yaxis_type="log",
        hovermode='closest',
        height=600,
        showlegend=True,
        legend=dict(
            orientation="h",
            yanchor="top",
            y=-0.2,
            xanchor="center",
            x=0.5
        )
    )
    
    # Ajouter une ligne horizontale à y=1 (dose sans écran)
    fig.add_hline(y=1, line_dash="dash", line_color="gray", 
                  annotation_text="Sans écran", annotation_position="right")
    
    return fig

print("✅ Fonction plot_normalized_dose créée !")

✅ Fonction plot_normalized_dose créée !


In [13]:
# D'abord, vérifions comment sont stockées les données "sans écran"
print("Valeurs uniques de Screen:")
print(data_wall['Screen'].unique())
print("\nValeurs uniques de Thickness pour Screen='None':")
none_screen = data_wall[data_wall['Screen'] == 'None']
if len(none_screen) > 0:
    print(none_screen['Thickness (cm)'].unique())
else:
    print("Aucune donnée avec Screen='None'")

Valeurs uniques de Screen:
['Concrete' 'Water' 'Lead' 'Steel' 'None']

Valeurs uniques de Thickness pour Screen='None':
[0]


In [20]:
def plot_attenuation_factor(data, x_variable='Thickness (cm)', filters=None):
    """
    Trace le facteur d'atténuation normalisé par rapport à la dose sans écran.
    Utilise les données avec Screen='None' et Thickness=0 comme référence.
    
    Parameters:
    -----------
    data : DataFrame
        Les données à tracer
    x_variable : str
        La variable à mettre en abscisse (généralement 'Thickness (cm)')
    filters : dict
        Dictionnaire de filtres à appliquer (Screen sera ignoré pour la référence)
    """
    
    # Copier les données
    df = data.copy()
    
    # Séparer les données avec et sans écran
    df_with_screen = df[df['Screen'] != 'None'].copy()
    df_no_screen = df[df['Screen'] == 'None'].copy()
    
    # Appliquer les filtres (sauf Screen)
    if filters:
        for column, values in filters.items():
            if column != 'Screen' and values:
                df_with_screen = df_with_screen[df_with_screen[column].isin(values)]
                df_no_screen = df_no_screen[df_no_screen[column].isin(values)]
    
    # Appliquer le filtre Screen uniquement sur df_with_screen
    if filters and 'Screen' in filters and filters['Screen']:
        df_with_screen = df_with_screen[df_with_screen['Screen'].isin(filters['Screen'])]
    
    # Créer le graphique
    fig = go.Figure()
    
    # Grouper par toutes les variables sauf x_variable
    group_columns = [col for col in ['Fissile', 'Case', 'Code', 'Screen', 'Particle', 'Distance (m)'] 
                     if col != x_variable and col in df_with_screen.columns]
    
    # Suivre les valeurs min/max pour l'échelle
    all_normalized_doses = []
    
    # Pour chaque groupe
    for name, group in df_with_screen.groupby(group_columns):
        # Créer un label descriptif
        if isinstance(name, tuple):
            label_dict = dict(zip(group_columns, name))
        else:
            label_dict = {group_columns[0]: name}
        
        # Créer le label sans Screen
        label_parts = [f"{col}={val}" for col, val in label_dict.items() if col != 'Screen']
        label_parts.insert(0, f"Screen={label_dict.get('Screen', 'Unknown')}")
        label = ", ".join(label_parts)
        
        # Trouver la dose de référence (sans écran) pour les mêmes conditions
        # (même Fissile, Case, Code, Particle, Distance)
        ref_filter = df_no_screen.copy()
        for col in ['Fissile', 'Case', 'Code', 'Particle']:
            if col in label_dict:
                ref_filter = ref_filter[ref_filter[col] == label_dict[col]]
        
        # Trier par x_variable
        group_sorted = group.sort_values(x_variable).copy()
        
        # Pour chaque point, normaliser par la dose de référence à la même distance
        normalized_doses = []
        normalized_uncertainties = []
        x_values = []
        
        for idx, row in group_sorted.iterrows():
            distance = row['Distance (m)']
            
            # Trouver la dose de référence à cette distance
            ref_at_distance = ref_filter[ref_filter['Distance (m)'] == distance]
            
            if len(ref_at_distance) > 0:
                ref_dose = ref_at_distance['Dose'].iloc[0]
                normalized_dose = row['Dose'] / ref_dose
                normalized_doses.append(normalized_dose)
                normalized_uncertainties.append(row['1s uncertainty'])
                x_values.append(row[x_variable])
                all_normalized_doses.append(normalized_dose)
        
        if len(normalized_doses) == 0:
            print(f"⚠️ Pas de données de référence pour {label}")
            continue
        
        # Ajouter la trace
        fig.add_trace(go.Scatter(
            x=x_values,
            y=normalized_doses,
            mode='lines+markers',
            name=label,
            error_y=dict(
                type='data',
                array=np.array(normalized_uncertainties) * np.array(normalized_doses),
                visible=True
            )
        ))
    
    # Calculer les limites de l'axe Y en échelle log
    if len(all_normalized_doses) > 0:
        min_dose = min(all_normalized_doses)
        # Limite inférieure : arrondir vers le bas en échelle log
        y_min = np.floor(np.log10(min_dose))
        # Limite supérieure : toujours 0 (10^0 = 1)
        y_max = 0
    else:
        y_min = -6
        y_max = 0
    
    # Mise en forme
    fig.update_layout(
        title=f"Facteur d'atténuation en fonction de {x_variable}<br>(Dose avec écran / Dose sans écran)",
        xaxis_title=x_variable,
        yaxis_title="Facteur d'atténuation",
        yaxis_type="log",
        yaxis_range=[y_min, y_max],  # Plage fixe
        hovermode='closest',
        height=600,
        showlegend=True,
        legend=dict(
            orientation="h",
            yanchor="top",
            y=-0.2,
            xanchor="center",
            x=0.5
        )
    )
    
    # Ajouter une ligne horizontale à y=1 (sans atténuation)
    fig.add_hline(y=1, line_dash="dash", line_color="gray", 
                  annotation_text="Sans écran", annotation_position="right")
    
    return fig

print("✅ Fonction plot_attenuation_factor créée !")

✅ Fonction plot_attenuation_factor créée !


In [21]:
# Graphique du facteur d'atténuation pour le plomb
filters = {
    'Particle': ['N'],
    'Screen': ['Lead'],  # Seulement le plomb
    'Case': ['C1 [U(4.95)O2F2 (H/235U = 410)]'],
    'Code': ['MCNP 6.1'],
}

fig = plot_attenuation_factor(data_wall, x_variable='Thickness (cm)', filters=filters)
fig.show()

### Comparaison des matériaux normalisés

In [22]:
# Comparer tous les matériaux à distance fixe (5m)
filters = {
    'Particle': ['N'],
    'Screen': ['Concrete', 'Steel', 'Lead', 'Water'],  # Tous les matériaux
    'Case': ['C1 [U(4.95)O2F2 (H/235U = 410)]'],
    'Code': ['MCNP 6.1'],
    'Distance (m)': [5]  # Distance fixe
}

fig = plot_attenuation_factor(data_wall, x_variable='Thickness (cm)', filters=filters)
fig.show()

## 📐 Extraction du coefficient d'atténuation μ

Fittons les données pour extraire μ pour chaque distance et chaque matériau d'écran

In [7]:
import scipy.optimize

def fit_mu_by_distance(data, screen='Lead', case=None, code=None, particle='N', model_type='simple'):
    """
    Fitte différents modèles d'atténuation pour chaque distance et extrait les paramètres.
    
    Models disponibles:
    - 'simple': D/D0 = exp(-mu * ep)
    - 'buildup': D/D0 = exp(-mu * ep) * (1 + B * ep)  # Facteur de build-up linéaire
    - 'double_exp': D/D0 = a * exp(-mu1 * ep) + (1-a) * exp(-mu2 * ep)  # Deux composantes
    - 'power': D/D0 = exp(-mu * ep^n)  # Loi de puissance
    """
    import numpy as np
    import pandas as pd
    import plotly.graph_objects as go
    
    # Filtrer les données
    df = data.copy()
    df_screen = df[df['Screen'] == screen]
    df_noscreen = df[df['Screen'] == 'None']
    if case:
        df_screen = df_screen[df_screen['Case'] == case]
        df_noscreen = df_noscreen[df_noscreen['Case'] == case]
    if code:
        df_screen = df_screen[df_screen['Code'] == code]
        df_noscreen = df_noscreen[df_noscreen['Code'] == code]
    if particle:
        df_screen = df_screen[df_screen['Particle'] == particle]
        df_noscreen = df_noscreen[df_noscreen['Particle'] == particle]
    
    # Définir les modèles
    def model_simple(ep, mu):
        return np.exp(-mu * ep)
    
    def model_buildup(ep, mu, B):
        """Avec facteur de build-up linéaire"""
        return np.exp(-mu * ep) * (1 + B * ep)
    
    def model_double_exp(ep, mu1, mu2, a):
        """Deux composantes exponentielles"""
        return a * np.exp(-mu1 * ep) + (1-a) * np.exp(-mu2 * ep)
    
    def model_power(ep, mu, n):
        """Loi de puissance dans l'exponentielle"""
        return np.exp(-mu * np.power(ep, n))
    
    # Choisir le modèle
    models = {
        'simple': (model_simple, ['mu'], [0.1]),
        'buildup': (model_buildup, ['mu', 'B'], [0.1, 0.01]),
        'double_exp': (model_double_exp, ['mu1', 'mu2', 'a'], [0.2, 0.05, 0.5]),
        'power': (model_power, ['mu', 'n'], [0.1, 1.0])
    }
    
    if model_type not in models:
        print(f"❌ Modèle '{model_type}' non reconnu. Choix: {list(models.keys())}")
        return None
    
    model_func, param_names, p0 = models[model_type]
    
    # Distances disponibles
    distances = sorted(df_screen['Distance (m)'].unique())
    results = []
    fig = go.Figure()
    
    for d in distances:
        # Données pour cette distance
        group = df_screen[df_screen['Distance (m)'] == d]
        ref = df_noscreen[df_noscreen['Distance (m)'] == d]
        if len(ref) == 0 or len(group) < len(p0):
            continue
        
        D0 = ref['Dose'].iloc[0]
        ep = group['Thickness (cm)'].values
        D = group['Dose'].values
        D_norm = D / D0
        unc = group['1s uncertainty'].values * D_norm
        
        # Fit
        try:
            popt, pcov = scipy.optimize.curve_fit(
                model_func, ep, D_norm, p0=p0, 
                sigma=unc, absolute_sigma=True,
                bounds=([0]*len(p0), [np.inf]*len(p0))  # Paramètres positifs
            )
            perr = np.sqrt(np.diag(pcov))
            
            # Stocker résultats
            res_dict = {'Distance (m)': d}
            for name, val, err in zip(param_names, popt, perr):
                res_dict[name] = val
                res_dict[f'{name}_err'] = err
            
            # Calculer R² pour évaluer la qualité du fit
            y_pred = model_func(ep, *popt)
            ss_res = np.sum((D_norm - y_pred)**2)
            ss_tot = np.sum((D_norm - np.mean(D_norm))**2)
            r2 = 1 - ss_res/ss_tot
            res_dict['R²'] = r2
            
            results.append(res_dict)
            
            # Tracer points et fit
            ep_fit = np.linspace(min(ep), max(ep), 100)
            D_fit = model_func(ep_fit, *popt)
            fig.add_trace(go.Scatter(x=ep, y=D_norm, mode='markers', 
                                    name=f'{d} m (data)', 
                                    error_y=dict(type='data', array=unc)))
            fig.add_trace(go.Scatter(x=ep_fit, y=D_fit, mode='lines', 
                                    name=f'{d} m (fit R²={r2:.3f})', 
                                    line=dict(dash='dash')))
        except Exception as e:
            print(f"⚠️ Fit échoué pour distance {d} m: {e}")
            continue
    
    fig.update_layout(
        title=f"Fit du facteur d'atténuation pour {screen}<br>Modèle: {model_type}",
        xaxis_title="Épaisseur (cm)",
        yaxis_title="D/D₀",
        yaxis_type="log",
        height=600,
        legend=dict(orientation="h", yanchor="top", y=-0.2, xanchor="center", x=0.5)
    )
    fig.show()
    
    # Résumé
    res_df = pd.DataFrame(results)
    print(f"\n📊 Résultats du fit ({model_type}):")
    print(res_df.to_string(index=False))
    return res_df

print("✅ Fonction fit_mu_by_distance créée avec plusieurs modèles !")
print("\nModèles disponibles:")
print("- 'simple': exp(-μ·ep)")
print("- 'buildup': exp(-μ·ep) × (1 + B·ep)")
print("- 'double_exp': a·exp(-μ₁·ep) + (1-a)·exp(-μ₂·ep)")
print("- 'power': exp(-μ·ep^n)")

✅ Fonction fit_mu_by_distance créée avec plusieurs modèles !

Modèles disponibles:
- 'simple': exp(-μ·ep)
- 'buildup': exp(-μ·ep) × (1 + B·ep)
- 'double_exp': a·exp(-μ₁·ep) + (1-a)·exp(-μ₂·ep)
- 'power': exp(-μ·ep^n)


### Test 1 : Modèle simple (exponentielle pure)

In [8]:
res_simple = fit_mu_by_distance(
    data_wall,
    screen='Lead',
    case='C1 [U(4.95)O2F2 (H/235U = 410)]',
    code='MCNP 6.1',
    particle='N',
    model_type='simple'
)


📊 Résultats du fit (simple):
 Distance (m)       mu   mu_err       R²
            1 0.028531 0.000283 0.957845
            2 0.031054 0.000383 0.988692
            5 0.036099 0.000379 0.980306
           10 0.037908 0.000310 0.982875
           20 0.039850 0.000222 0.996911
           50 0.041211 0.000202 0.994977
          100 0.041906 0.000266 0.932467
          200 0.041655 0.000172 0.999066
          300 0.042166 0.000189 0.996177
          500 0.045522 0.000166 0.987388
          700 0.049661 0.000159 0.978398
         1000 0.055708 0.000119 0.955302
         1200 0.058482 0.000124 0.954619


### Test 2 : Modèle avec build-up

Le facteur de build-up représente l'accumulation de rayonnement diffusé

In [9]:
res_buildup = fit_mu_by_distance(
    data_wall,
    screen='Lead',
    case='C1 [U(4.95)O2F2 (H/235U = 410)]',
    code='MCNP 6.1',
    particle='N',
    model_type='buildup'
)


📊 Résultats du fit (buildup):
 Distance (m)       mu       mu_err            B        B_err       R²
            1 0.061293     0.002540 4.500634e-02     0.004592 0.989947
            2 0.054849     0.004318 2.980729e-02     0.006621 0.996597
            5 0.070720     0.003331 4.865253e-02     0.006272 0.999203
           10 0.045576     0.012978 8.253957e-03     0.015003 0.983242
           20 0.062267     0.003050 2.788413e-02     0.004626 0.999146
           50 0.041211 10664.578918 1.188327e-09 10664.577644 0.994977
          100 0.041906  2518.051152 4.069760e-08  2518.052741 0.932467
          200 0.041655 21497.205937 4.874806e-09 21497.204876 0.999066
          300 0.047023     0.014532 5.092052e-03     0.015959 0.996250
          500 0.045522 19953.615686 5.103903e-09 19953.614791 0.987388
          700 0.049661  6679.612991 7.945458e-10  6679.612168 0.978398
         1000 0.055708  5511.532518 5.810685e-10  5511.531798 0.955302
         1200 0.090333     0.001308 4.373396e-

### Test 3 : Modèle double exponentielle

Utile si le spectre a deux composantes dominantes (ex: neutrons rapides + thermiques)

In [10]:
res_double = fit_mu_by_distance(
    data_wall,
    screen='Lead',
    case='C1 [U(4.95)O2F2 (H/235U = 410)]',
    code='MCNP 6.1',
    particle='N',
    model_type='double_exp'
)


📊 Résultats du fit (double_exp):
 Distance (m)      mu1      mu1_err          mu2    mu2_err        a        a_err       R²
            1 0.028540    69.116343 2.852920e-02   9.283735 0.158505 9.884364e+05 0.957845
            2 0.031054   944.828709 3.105452e-02 191.516432 0.167053 2.556792e+06 0.988692
            5 0.036100   460.017586 3.609882e-02 108.990641 0.201268 6.988573e+06 0.980306
           10 0.037896   132.981419 3.791130e-02  48.965982 0.232802 4.532722e+06 0.982875
           20 0.039870 24082.244249 3.984969e-02  17.824725 0.000558 1.470866e+06 0.996911
           50 0.053977     0.084024 8.151264e-15   0.277102 0.847166 2.014982e+00 0.998174
          100 0.164890     0.104132 2.546004e-02   0.011441 0.285148 2.124218e-01 0.988250
          200 0.160334     0.892022 4.021006e-02   0.008942 0.029784 2.429318e-01 0.999274
          300 0.042163   120.509238 4.216680e-02  26.139371 0.250987 8.325241e+06 0.996177
          500 0.218564     0.224622 4.192671e-02   0.003

### Test 4 : Modèle loi de puissance

Teste si l'atténuation suit exp(-μ·ep^n) au lieu de exp(-μ·ep)

In [11]:
res_power = fit_mu_by_distance(
    data_wall,
    screen='Lead',
    case='C1 [U(4.95)O2F2 (H/235U = 410)]',
    code='MCNP 6.1',
    particle='N',
    model_type='power'
)


📊 Résultats du fit (power):
 Distance (m)       mu   mu_err        n    n_err       R²
            1 0.013286 0.001732 1.262677 0.044398 0.992459
            2 0.021915 0.002872 1.120684 0.044993 0.996745
            5 0.018297 0.002590 1.232642 0.048106 0.998633
           10 0.035993 0.003663 1.017725 0.034676 0.984059
           20 0.031565 0.002148 1.079808 0.023189 0.998549
           50 0.051647 0.003041 0.922867 0.020078 0.998590
          100 0.087088 0.005550 0.749054 0.021879 0.995889
          200 0.044601 0.002298 0.976643 0.017560 0.999144
          300 0.040564 0.002535 1.013149 0.021158 0.996527
          500 0.055154 0.002079 0.933945 0.012935 0.986840
          700 0.055999 0.001961 0.958474 0.012063 0.972380
         1000 0.055833 0.001511 0.999231 0.009219 0.955198
         1200 0.041117 0.001210 1.120050 0.009983 0.971636


### 📊 Comparaison des modèles

Comparons les R² pour voir quel modèle fitte le mieux les données

In [13]:
# Comparer les R² moyens de chaque modèle
import pandas as pd

comparison = pd.DataFrame({
    'Distance (m)': res_simple['Distance (m)'],
    'R² Simple': res_simple['R²'],
    'R² Buildup': res_buildup['R²'],
    'R² Double_exp': res_double['R²'],
    'R² Power': res_power['R²']
})

print("📊 Comparaison des R² par distance:")
print(comparison.to_string(index=False))

print("\n📈 R² moyens:")
print(f"Simple (exp(-μ·ep)):           {res_simple['R²'].mean():.6f}")
print(f"Buildup (exp(-μ·ep)×(1+B·ep)): {res_buildup['R²'].mean():.6f}")
print(f"Double exp:                     {res_double['R²'].mean():.6f}")
print(f"Power (exp(-μ·ep^n)):          {res_power['R²'].mean():.6f}")

# Tracer la comparaison
import plotly.graph_objects as go
fig = go.Figure()
fig.add_trace(go.Scatter(x=comparison['Distance (m)'], y=comparison['R² Simple'], 
                         mode='lines+markers', name='Simple'))
fig.add_trace(go.Scatter(x=comparison['Distance (m)'], y=comparison['R² Buildup'], 
                         mode='lines+markers', name='Buildup'))
fig.add_trace(go.Scatter(x=comparison['Distance (m)'], y=comparison['R² Double_exp'], 
                         mode='lines+markers', name='Double exp'))
fig.add_trace(go.Scatter(x=comparison['Distance (m)'], y=comparison['R² Power'], 
                         mode='lines+markers', name='Power'))
fig.update_layout(title="Comparaison de la qualité des fits (R²)",
                  xaxis_title="Distance (m)", 
                  xaxis_type="log",
                  yaxis_title="R²",
                  yaxis_range=[0.9, 1.0],
                  height=500)
fig.show()

📊 Comparaison des R² par distance:
 Distance (m)  R² Simple  R² Buildup  R² Double_exp  R² Power
            1   0.957845    0.989947       0.957845  0.992459
            2   0.988692    0.996597       0.988692  0.996745
            5   0.980306    0.999203       0.980306  0.998633
           10   0.982875    0.983242       0.982875  0.984059
           20   0.996911    0.999146       0.996911  0.998549
           50   0.994977    0.994977       0.998174  0.998590
          100   0.932467    0.932467       0.988250  0.995889
          200   0.999066    0.999066       0.999274  0.999144
          300   0.996177    0.996250       0.996177  0.996527
          500   0.987388    0.987388       0.988109  0.986840
          700   0.978398    0.978398       0.974021  0.972380
         1000   0.955302    0.955302       0.954661  0.955198
         1200   0.954619    0.967494       0.954619  0.971636

📈 R² moyens:
Simple (exp(-μ·ep)):           0.977309
Buildup (exp(-μ·ep)×(1+B·ep)): 0.983037
Dou

## 🎯 Conclusions sur les modèles d'atténuation

### Résumé des R² moyens :
1. **Power (exp(-μ·ep^n))** : **R² = 0.9882** ✨ Meilleur modèle
2. **Buildup (exp(-μ·ep)×(1+B·ep))** : R² = 0.9830
3. **Double exp** : R² = 0.9815
4. **Simple (exp(-μ·ep))** : R² = 0.9773

### 📝 Interprétation :

**Modèle Power (meilleur fit)** :
- Le paramètre **n** varie autour de **1.0** (entre 0.75 et 1.26)
- Quand n ≈ 1 → équivalent au modèle simple
- Quand n < 1 → atténuation sous-linéaire (build-up possible)
- Quand n > 1 → atténuation sur-linéaire (effet de durcissement du spectre)

**Variation de μ avec la distance** :
- Pour le modèle simple : μ augmente de **0.029 cm⁻¹** (1m) à **0.058 cm⁻¹** (1200m)
- Cette variation suggère que le spectre change avec la distance
- À grande distance : neutrons plus thermalisés → section efficace plus élevée

**Recommandations** :
- Pour une analyse simple : utiliser le **modèle simple** avec μ dépendant de la distance
- Pour une meilleure précision : utiliser le **modèle Power** qui capture les effets spectraux
- Le coefficient d'atténuation **n'est pas constant** → dépend de la distance (effet spectral)